src/signing-keys.asc

```diff
@@ -0,0 +1,91 @@
+-----BEGIN PGP PUBLIC KEY BLOCK-----
```

Comment:
Hmm, why not just pick up the keys from `fedora-gpg-keys`/`redhat-release-coreos` (i.e. `/etc/pki/rpm-gpg`) ?

Reply:
coreos-installer should run on arbitrary distros, which may not have those files.

Reply:
OK so the idea here is that we'll publish binaries instead of/in addition to containers? What I'm trying to avoid is adding yet another spot that needs updating during each major release rollover. Can we pull it in as part of the build process at least?

Reply:
It may make sense to overlay-merge (by filename?) two sets of keys:
 * these hardcoded ones (for fallback)
 * any on-disk ones (possibly fresher)

Reply:
> OK so the idea here is that we'll publish binaries instead of/in addition to containers?

It was, but if the container works and the UX for running it isn't too bad, I'd favor shipping that instead of a binary.

> any on-disk ones (possibly fresher)

Where would we get these?  If from `/etc/pki/rpm-gpg`, there are no symbolic references for the current key, so we'd have to either import everything we find there (boo) or have a heuristic for importing keys (boo).

If we do ship a bare binary, this would also mean that behavior would differ depending on the host OS, which doesn't seem great.

> What I'm trying to avoid is adding yet another spot that needs updating during each major release rollover. Can we pull it in as part of the build process at least?

These are the roots of trust for install images, and IMO it's safer to manage them manually.  We can always start that way and add dynamic roots or more automation later on.

In comparison, CL `coreos-install` intentionally hardcodes the GPG key and does not provide a command-line mechanism for overriding it.

Reply:
So as an example right now we don't have any CI covering the `coreos-installer` path. So bumping from f30 to f31 without bumping the keys here would've possibly gone completely unnoticed until it shipped to users. (A gap we should fix of course!)

> Where would we get these? If from /etc/pki/rpm-gpg, there are no symbolic references for the current key, so we'd have to either import everything we find there (boo) or have a heuristic for importing keys (boo).

> These are the roots of trust for install images, and IMO it's safer to manage them manually. We can always start that way and add dynamic roots or more automation later on.

A big part of this discussion is essentially a direct consequence of https://github.com/coreos/fedora-coreos-tracker/issues/187#issuecomment-540103433. WDYT about:
- ship in a container with `fedora-gpg-keys`
- parse the FCOS version to extract the major release, similar to what is proposed for RoboSignatory itself in that ticket
- verify download using *only that release's specific key* from `/etc/pki/rpm-gpg`

?

This is similar to how Fedora verifies RPMs: https://src.fedoraproject.org/rpms/fedora-repos/blob/a12c809ab2dcb730cbeb4a03c50d20e3d25cc048/f/fedora.repo#_10.

For RHEL, the keys don't rotate nearly as much, so we can just bake it in.

Anyway, I'm cool with how things set up for now, and discussing this stuff in more details later on!

Reply:
> So as an example right now we don't have any CI covering the coreos-installer path. So bumping from f30 to f31 without bumping the keys here would've possibly gone completely unnoticed until it shipped to users. (A gap we should fix of course!)

If nothing else, that'll be addressed when we start running kola on Packet, where every machine will bootstrap with coreos-installer.

> ship in a container with fedora-gpg-keys

I'd still prefer not to rely on external keys.  It'd be great if `cargo install coreos-installer` worked properly on e.g. Debian.

>  parse the FCOS version to extract the major release, similar to what is proposed for RoboSignatory itself in that ticket
> verify download using only that release's specific key from /etc/pki/rpm-gpg

I like the idea of binding to a specific key.  Will file an issue.

Hmm, maybe stream metadata should list a signing key ID for each artifact?

Reply:
https://github.com/coreos/coreos-installer/issues/71